include/kernels/coarse_op_kernel.cuh

```diff
@@ -687,6 +918,110 @@ namespace quda {
+     Where: s = fine spin, c' = coarse color, c = fine color, d = hypercube corner
+     See staggered_kd_apply_xinv_kernel.cuh for reference
+
+     If Arg::compute_max is true, then we do not save the result, and
+     instead merely compute the maximum element.  This is used for
+     setting the scale for fixed point.
+  */
+  template <typename Arg> struct compute_kv {
+    using real = typename Arg::Float;
+    using Vector = ColorSpinor<real, Arg::fineColor, 1>;
+    using Link = Matrix<complex<float>, Arg::fineColor>;
+
+    const Arg &arg;
+    static constexpr const char *filename() { return KERNEL_FILE; }
+    constexpr compute_kv(const Arg &arg) : arg(arg) { }
+
+    /**
+       3-d parallelism
+       @param[in] x_cb e/o fine-grid spacetime
+       @param[in] ch_parity [chirality (clover), source parity (staggered)] and parity
+       @param[in] c_row output color column
+    */
+    __device__ __host__ inline void operator()(int x_cb, int ch_parity, int ic_c)
+    {
+      const int nbr_parity = ch_parity % 2;
+      const int parity = ch_parity / 2;
+
+      // Get coordinates
+      constexpr auto nDim = 4;
+      Coord<nDim> coord;
+      coord.x_cb = x_cb;
+      coord.X = getCoordsCB(coord, x_cb, arg.x_size, arg.x_size[0] / 2, parity);
+
+      // Get location of unit corner of hypercube
+      int x_c[nDim];
+#pragma unroll
+      for (int d = 0; d < nDim; d++)
+        x_c[d] = 2 * (coord[d] / 2);
+
+      Vector out;
+
+      // only needed for dagger
+      // global parity == parity w/in the KD block
+      int my_corner = 8 * parity + 4 * (coord[3] % 2) + 2 * (coord[2] % 2) + (coord[1] % 2);
+
+      // Begin accumulating into the output vector
+      // start at 0 -> store in spin 0 (even source)
+      // start at 8 -> store in spin 1 (odd source)
+      int nbr_corner = (nbr_parity == 0) ? 0 : 8;
+
+#pragma unroll
+      for (int nbr_t = 0; nbr_t < 2; nbr_t++) {
+#pragma unroll
+        for (int nbr_z = 0; nbr_z < 2; nbr_z++) {
+#pragma unroll
+          for (int nbr_y = 0; nbr_y < 2; nbr_y++) {
+            const int offset[nDim] = { (nbr_parity + nbr_t + nbr_z + nbr_y) & 1, nbr_y, nbr_z, nbr_t };
+            const int nbr_x_cb = linkIndexShift(x_c, offset, arg.x_size);
+
+            // Load Xinv link
+            Link Xinv;
+#pragma unroll
+            for (int ic_f = 0; ic_f < Arg::fineColor; ic_f++) {
+#pragma unroll
+              for (int jc_f = 0; jc_f < Arg::fineColor; jc_f++) {
+                Xinv(ic_f, jc_f) = arg.kd_dagger ? arg.K(my_corner, nbr_parity, nbr_x_cb, 0, 0, ic_f, jc_f) : arg.K(nbr_corner, parity, x_cb, 0, 0, ic_f, jc_f);
+              }
+            }
+
+            // Load spinor
+            Vector in;
+#pragma unroll
+            for (int jc_f = 0; jc_f < Arg::fineColor; jc_f++) {
+              in(0, jc_f) = arg.V(nbr_parity, nbr_x_cb, 0, jc_f, ic_c);
+            }
+
+            out = mv_add(arg.kd_dagger ? conj(Xinv) : Xinv, in, out);
```

Comment:
This expression to my eye looks like it could cause the compiler to spill.  Am I being paranoid?

Reply:
I haven't seen any compiler complaints on spilling for this kernel or for the kernel it's based on in `staggered_kd_apply_xinv.cu`, so I think (perhaps by good graces) it's safe.